File: sdk/go/krasis/notes.go
package krasis

import (
	"fmt"
	"net/url"
)

// NotesModule handles note CRUD operations.
type NotesModule struct {
	client *Client
}

// NewNotesModule creates a new notes module.
func NewNotesModule(c *Client) *NotesModule {
	return &NotesModule{client: c}
}

// ListNotesOptions for filtering notes.
type ListNotesOptions struct {
	FolderID string
	Page     int
	Size     int
}

// List retrieves a paginated list of notes.
func (n *NotesModule) List(opts *ListNotesOptions) (*Paginated[Note], error) {
	if opts == nil {
		opts = &ListNotesOptions{Page: 1, Size: 20}
	}
	params := url.Values{}
	params.Set("page", fmt.Sprintf("%d", opts.Page))
	params.Set("size", fmt.Sprintf("%d", opts.Size))
	if opts.FolderID != "" {
		params.Set("folder_id", opts.FolderID)
	}
	var result Paginated[Note]
	err := n.client.Get("/notes?"+params.Encode(), &result)
	return &result, err
}

// Create creates a new note.
func (n *NotesModule) Create(title, content string, folderID *string) (*Note, error) {
	body := map[string]any{"title": title, "content": content}
	if folderID != nil {
		body["folder_id"] = *folderID
	}
	var note Note
	err := n.client.Post("/notes", body, &note)
	return &note, err
}

// Get retrieves a note by ID.
func (n *NotesModule) Get(id string) (*Note, error) {
	var note Note
	err := n.client.Get("/notes/"+id, &note)
	return &note, err
}

// UpdateNoteOptions for updating a note.
type UpdateNoteOptions struct {
	Title         *string
	Content       *string
	FolderID      *string
	IsPublic      *bool
	Version       int
	ChangeSummary *string
}

// Update modifies an existing note.
func (n *NotesModule) Update(id string, opts UpdateNoteOptions) (*Note, error) {
	body := map[string]any{}
	if opts.Title != nil {
		body["title"] = *opts.Title
	}
	if opts.Content != nil {
		body["content"] = *opts.Content
	}
	if opts.FolderID != nil {
		body["folder_id"] = *opts.FolderID
	}
	if opts.IsPublic != nil {
		body["is_public"] = *opts.IsPublic
	}
	if opts.ChangeSummary != nil {
		body["change_summary"] = *opts.ChangeSummary
	}

	headers := map[string]string{}
	if opts.Version > 0 {
		headers["If-Match"] = fmt.Sprintf("%d", opts.Version)
	}

	var note Note
	err := n.client.Put("/notes/"+id, body, &note, headers)
	return &note, err
}

// Delete removes a note.
func (n *NotesModule) Delete(id string) error {
	return n.client.Delete("/notes/"+id, nil)
}

// Versions returns the version history of a note.
func (n *NotesModule) Versions(id string) ([]NoteVersion, error) {
	var versions []NoteVersion
	err := n.client.Get("/notes/"+id+"/versions", &versions)
	return versions, err
}

// RestoreVersion rolls back a note to a previous version.
func (n *NotesModule) RestoreVersion(id string, version int) (*Note, error) {
	var note Note
	err := n.client.Post(fmt.Sprintf("/notes/%s/versions/%d/restore", id, version), nil, &note)
	return &note, err
}

// FoldersModule handles folder CRUD operations.
type FoldersModule struct {
	client *Client
}

// NewFoldersModule creates a new folders module.
func NewFoldersModule(c *Client) *FoldersModule {
	return &FoldersModule{client: c}
}

// List returns all folders.
func (f *FoldersModule) List() ([]Folder, error) {
	var folders []Folder
	err := f.client.Get("/folders", &folders)
	return folders, err
}

// Create creates a new folder.
func (f *FoldersModule) Create(name string, parentID, color *string, sortOrder int) (*Folder, error) {
	body := map[string]any{"name": name, "sort_order": sortOrder}
	if parentID != nil {
		body["parent_id"] = *parentID
	}
	if color != nil {
		body["color"] = *color
	}
	var folder Folder
	err := f.client.Post("/folders", body, &folder)
	return &folder, err
}

// Update modifies an existing folder.
func (f *FoldersModule) Update(id string, name string, parentID, color *string, sortOrder *int) (*Folder, error) {
	body := map[string]any{}
	if name != "" {
		body["name"] = name
	}
	if parentID != nil {
		body["parent_id"] = *parentID
	}
	if color != nil {
		body["color"] = *color
	}
	if sortOrder != nil {
		body["sort_order"] = *sortOrder
	}
	var folder Folder
	err := f.client.Put("/folders/"+id, body, &folder, nil)
	return &folder, err
}

// Delete removes a folder.
func (f *FoldersModule) Delete(id string) error {
	return f.client.Delete("/folders/"+id, nil)
}

// ShareModule handles note sharing operations.
type ShareModule struct {
	client *Client
}

// NewShareModule creates a new share module.
func NewShareModule(c *Client) *ShareModule {
	return &ShareModule{client: c}
}

// CreateShareOptions for creating a share.
type CreateShareOptions struct {
	Permission string
	Password   string
	ExpiresAt  string
}

// Create generates a share for a note.
func (s *ShareModule) Create(noteID string, opts *CreateShareOptions) (*ShareStatus, error) {
	body := map[string]any{"permission": "read"}
	if opts != nil {
		if opts.Permission != "" {
			body["permission"] = opts.Permission
		}
		if opts.Password != "" {
			body["password"] = opts.Password
		}
		if opts.ExpiresAt != "" {
			body["expires_at"] = opts.ExpiresAt
		}
	}
	var share ShareStatus
	err := s.client.Post("/notes/"+noteID+"/share", body, &share)
	return &share, err
}

// Get returns the share status for a note.
func (s *ShareModule) Get(noteID string) (*ShareStatus, error) {
	var share ShareStatus
	err := s.client.Get("/notes/"+noteID+"/share", &share)
	return &share, err
}

// Revoke deletes an existing share.
func (s *ShareModule) Revoke(noteID string) error {
	return s.client.Delete("/notes/"+noteID+"/share", nil)
}

// AccessByToken accesses a shared note by token.
func (s *ShareModule) AccessByToken(token string, password string) (*Note, error) {
	path := "/share/" + token
	if password != "" {
		path += "?password=" + url.QueryEscape(password)
	}
	var note Note
	err := s.client.Get(path, &note)
	return &note, err
}

File: sdk/python/krasis/notes.py
"""Notes, folders, and sharing modules."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from krasis.client import Client
from krasis.types import Folder, Note, NoteVersion, Paginated, ShareStatus


@dataclass
class ListNotesOptions:
    folder_id: str = ""
    page: int = 1
    size: int = 20


@dataclass
class UpdateNoteOptions:
    title: str | None = None
    content: str | None = None
    folder_id: str | None = None
    is_public: bool | None = None
    version: int = 0
    change_summary: str | None = None


class NotesModule:
    """Note CRUD operations."""

    def __init__(self, client: Client) -> None:
        self._client = client

    async def list(self, opts: ListNotesOptions | None = None) -> Paginated[Note]:
        if opts is None:
            opts = ListNotesOptions()
        params: dict[str, str] = {
            "page": str(opts.page),
            "size": str(opts.size),
        }
        if opts.folder_id:
            params["folder_id"] = opts.folder_id
        qs = "&".join(f"{k}={v}" for k, v in params.items())
        return await self._client.get_paginated(f"/notes?{qs}", Note)

    async def create(
        self, title: str, content: str = "", folder_id: str | None = None
    ) -> Note:
        body: dict[str, Any] = {"title": title, "content": content}
        if folder_id:
            body["folder_id"] = folder_id
        data = await self._client.post("/notes", body=body)
        return Note(**data)

    async def get(self, note_id: str) -> Note:
        data = await self._client.get(f"/notes/{note_id}")
        return Note(**data)

    async def update(self, note_id: str, opts: UpdateNoteOptions) -> Note:
        body: dict[str, Any] = {}
        if opts.title is not None:
            body["title"] = opts.title
        if opts.content is not None:
            body["content"] = opts.content
        if opts.folder_id is not None:
            body["folder_id"] = opts.folder_id
        if opts.is_public is not None:
            body["is_public"] = opts.is_public
        if opts.change_summary is not None:
            body["change_summary"] = opts.change_summary

        headers: dict[str, str] = {}
        if opts.version > 0:
            headers["If-Match"] = str(opts.version)

        data = await self._client.put(f"/notes/{note_id}", body=body, headers=headers)
        return Note(**data)

    async def delete(self, note_id: str) -> None:
        await self._client.delete(f"/notes/{note_id}")

    async def versions(self, note_id: str) -> list[NoteVersion]:
        data = await self._client.get(f"/notes/{note_id}/versions")
        return [NoteVersion(**v) for v in data]

    async def restore_version(self, note_id: str, version: int) -> Note:
        data = await self._client.post(f"/notes/{note_id}/versions/{version}/restore")
        return Note(**data)


class FoldersModule:
    """Folder CRUD operations."""

    def __init__(self, client: Client) -> None:
        self._client = client

    async def list(self) -> list[Folder]:
        data = await self._client.get("/folders")
        return [Folder(**f) for f in data]

    async def create(
        self, name: str, parent_id: str | None = None, color: str | None = None, sort_order: int = 0
    ) -> Folder:
        body: dict[str, Any] = {"name": name, "sort_order": sort_order}
        if parent_id:
            body["parent_id"] = parent_id
        if color:
            body["color"] = color
        data = await self._client.post("/folders", body=body)
        return Folder(**data)

    async def update(
        self,
        folder_id: str,
        name: str = "",
        parent_id: str | None = None,
        color: str | None = None,
        sort_order: int | None = None,
    ) -> Folder:
        body: dict[str, Any] = {}
        if name:
            body["name"] = name
        if parent_id:
            body["parent_id"] = parent_id
        if color:
            body["color"] = color
        if sort_order is not None:
            body["sort_order"] = sort_order
        data = await self._client.put(f"/folders/{folder_id}", body=body)
        return Folder(**data)

    async def delete(self, folder_id: str) -> None:
        await self._client.delete(f"/folders/{folder_id}")


class ShareModule:
    """Note sharing operations."""

    def __init__(self, client: Client) -> None:
        self._client = client

    async def create(
        self,
        note_id: str,
        permission: str = "read",
        password: str = "",
        expires_at: str = "",
    ) -> ShareStatus:
        body: dict[str, Any] = {"permission": permission}
        if password:
            body["password"] = password
        if expires_at:
            body["expires_at"] = expires_at
        data = await self._client.post(f"/notes/{note_id}/share", body=body)
        return ShareStatus(**data)

    async def get(self, note_id: str) -> ShareStatus:
        data = await self._client.get(f"/notes/{note_id}/share")
        return ShareStatus(**data)

    async def revoke(self, note_id: str) -> None:
        await self._client.delete(f"/notes/{note_id}/share")

    async def access_by_token(self, token: str, password: str = "") -> Note:
        path = f"/share/{token}"
        if password:
            from urllib.parse import quote
            path += f"?password={quote(password)}"
        data = await self._client.get(path)
        return Note(**data)
